From 150a879224353bfa825af4f222e7eaacda3b3843 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Thu, 28 Mar 2019 23:23:20 -0700
Subject: [PATCH 0588/1851] v4l: xilinx: tpg: Dynamically modify TPG output
 format

commit d5579a60c068d3ffb2bc38d73249b8de234e1ff8 from
https://github.com/Xilinx/linux-xlnx.git

The TPG output format was fixed based on the device tree. This patch now
allows the TPG output format to be configured at runtime. This is valid
only for the HLS version of TPG.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Tested-by: Rohan Dhaval Parikh <rohanp@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-tpg.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/media/platform/xilinx/xilinx-tpg.c b/drivers/media/platform/xilinx/xilinx-tpg.c
index 1fa5554f468a..f840bc098d9e 100644
--- a/drivers/media/platform/xilinx/xilinx-tpg.c
+++ b/drivers/media/platform/xilinx/xilinx-tpg.c
@@ -306,7 +306,7 @@ static int xtpg_s_stream(struct v4l2_subdev *subdev, int enable)
 	if (xtpg->is_hls) {
 		u32 fmt = 0;
 
-		switch (xtpg->vip_format->code) {
+		switch (xtpg->formats[0].code) {
 		case MEDIA_BUS_FMT_VYYUYY8_1X24:
 		case MEDIA_BUS_FMT_VYYUYY10_4X20:
 			fmt = XTPG_HLS_COLOR_FORMAT_YUV_420;
@@ -438,6 +438,23 @@ static int xtpg_set_format(struct v4l2_subdev *subdev,
 			__format->code = fmt->format.code;
 	}
 
+	if (xtpg->is_hls) {
+		switch (fmt->format.code) {
+		case MEDIA_BUS_FMT_VYYUYY8_1X24:
+		case MEDIA_BUS_FMT_VYYUYY10_4X20:
+		case MEDIA_BUS_FMT_UYVY8_1X16:
+		case MEDIA_BUS_FMT_UYVY10_1X20:
+		case MEDIA_BUS_FMT_VUY8_1X24:
+		case MEDIA_BUS_FMT_VUY10_1X30:
+		case MEDIA_BUS_FMT_RBG888_1X24:
+		case MEDIA_BUS_FMT_RBG101010_1X30:
+			__format->code = fmt->format.code;
+			break;
+		default:
+			__format->code = xtpg->default_format.code;
+		}
+	}
+
 	__format->width = clamp_t(unsigned int, fmt->format.width,
 				  XTPG_MIN_WIDTH, xtpg->max_width);
 	__format->height = clamp_t(unsigned int, fmt->format.height,
-- 
2.31.1

